src/Avalonia.Base/StyledElement.cs

```diff
@@ -930,38 +901,24 @@ private void DetachStylesFromThisAndDescendents(IReadOnlyList<IStyle> styles)
             }
         }
 
-        private static IReadOnlyList<IStyle> RecurseStyles(IReadOnlyList<IStyle> styles)
+        private static IReadOnlyList<StyleBase> RecurseStyles(IReadOnlyList<IStyle> styles)
         {
-            var count = styles.Count;
-            List<IStyle>? result = null;
-
-            for (var i = 0; i < count; ++i)
-            {
-                var style = styles[i];
-
-                if (style.Children.Count > 0)
-                {
-                    if (result is null)
-                    {
-                        result = new List<IStyle>(styles);
-                    }
-
-                    RecurseStyles(style.Children, result);
-                }
-            }
-
-            return result ?? styles;
+            var result = new List<StyleBase>();
+            RecurseStyles(styles, result);
+            return result;
         }
 
-        private static void RecurseStyles(IReadOnlyList<IStyle> styles, List<IStyle> result)
+        private static void RecurseStyles(IReadOnlyList<IStyle> styles, List<StyleBase> result)
         {
             var count = styles.Count;
 
             for (var i = 0; i < count; ++i)
             {
-                var style = styles[i];
-                result.Add(style);
-                RecurseStyles(style.Children, result);
+                var s = styles[i];
+                if (s is StyleBase style)
```

Comment:
Maybe it would make sense to add a method to `IStyle` instead of using reflection data (type id).

Reply:
I need to refactor how styles are applied anyway, this was the simplest way to make this code work currently.